Fix state not being shared with "beforeXXX" and "afterXXX" events

rgoupil opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the documentation. (Should be made against the documentation branch)
- Create or update the tests.
- Refer to the issue you are closing in the PR description - fix #issue
- Specify if the PR is in WIP (work in progress) state or ready to be merged

Please ensure you read through the Contributing Guide: https://github.com/strapi/strapi/blob/master/CONTRIBUTING.md
-->

### What does it do?

Fix state not being shared between "beforeXXX" and "afterXXX" events, as described in the documentation here:
https://docs.strapi.io/developer-docs/latest/development/backend-customization/models.html#hook-event-object

### Why is it needed?

The feature is described in the documentation.
Sharing state can be useful to update external resources for instance.

### How to test it?

* Create a content-type `<type>` from the UI
* Create `lifecycles.js` file in `src/api/<type>/content-types/<type>`
* Minimal test case code
```js

module.exports = {
    async beforeUpdate(event) {
        event.state = 'hello world';
        return event;
    },

    async afterUpdate(event) {
        // event.state is undefined before this fix, but it should be set to 'hello world'
        console.log(event.state);
    },
};
```

### Related issue(s)/PR(s)

None

alexandrebodin commented on packages/core/database/lib/lifecycles/index.js:
```diff
@@ -30,31 +30,39 @@ const createLifecyclesProvider = db => {
 
-    async run(action, uid, properties) {
-      for (const subscriber of subscribers) {
+    async run(action, uid, properties, states) {
```
I don't understand the logic you use with states being considered as an array here? 
We want to pass the same state to all subscribers just passing a simple state object around would be enough I think

WalkingPizza replied:
Sorry to intrude, but I have a question.
Is the `state` object shared across all lifecycles or is it model-specific?

rgoupil replied:
From my understanding, using only one state object create the risk of a subscriber replacing the values put in the state by a previous subscriber, potentially causing issues for future events down the line.

Say we plug in two external libraries along with a couple of custom subscribers, I now have their states mixed with mine and I am at risk of replacing critical values by mistake. Of course this will most likely break the code of the events registered by the external libraries as the state doen't match their expectation anymore.

Using a different state for each subscriber identity prevent this issue and all events receive exactly the state they defined, no matter how many subscribers are registered.

I'll be happy to be proven wrong and to use a simple object rather than an array of objects, please let me know!

alexandrebodin replied:
Hi @rgoupil thank you for elaborating on the idea you had. It does make a lot of sense to scope the states per subscriber indeed 👍 I am still thinking a global state could be very useful but more for a middleware-like approach though. we could introduce that in the future.

That being said, I would still pass around a single "state" even if it contains some logic to distribute the right context to each subscriber. I think the lifecycle should be more "dumb" in that regard. Here is what I have in mind

```
const { state } = await db.lifecycles.run('beforeFindOne');
//...
 await db.lifecycles.run('afterFindOne', ..., state);
```

Where state would be a Map from subscriber to state. This would avoid race conditions that the array index approach can have.

Each lifecycle could return its state and initialize one if none was provided. Pretty close to your approach but more abstracted away from the entity manager.

Hope this makes sense :)

rgoupil replied:
It makes total sense, I'll do just that :)
I disliked the `Array` approach however I failed to remember that a `Map` can handle objects and functions as keys (as opposed to an `Object`), thanks for the reminder!

rgoupil replied:
I updated the implementation to reflect your changes.
Thanks for the review, I'll mark this conversation as resolved!

WalkingPizza commented:
Just implemented the changes requested in #12834.

I hope my submission of a competing PR does not offend you @rgoupil.
I am relying on this functionality for a project I have in production and would like to see the fix merged asap, hence the PR.

rgoupil commented:
Hey @WalkingPizza thanks for the heads up. We have been using my PR in production for a few months now, I understand your situation no offense taken ;)

I answered the comment as I believe using a single object instead of an array goes against the original v3 documentation and will cause issues and lead to a barely useful feature.

Feel free to read my answer and pitch in :)

rgoupil commented:
Hey @alexandrebodin, I was wondering what did you though of my answer about your requested change? Thanks!

Eggwise commented:
@alexandrebodin Looking forward for this fix.

Defining lifecycle hooks programaticly currently is not working

alexandrebodin commented:
Hi @rgoupil  will try to have a look this week 🙏 

@Eggwise Can you elaborate? this PR adds a state feature on top of lifecycles but they are already working. What is the issue you are facing ?

alexandrebodin reviewed (commented):
Everything looks fine :) Can you add some unit testing on the lifecycle code to validate state management is working as expected ?

PritamSangani commented:
@rgoupil - any updates on when you think you will be able to add the unit tests to complete the PR?

rgoupil commented:
@alexandrebodin I added a couple unit tests to cover the `Lifecycles.run` method.

Please also have a look at 0f0b030, as I fixed the wrong `after` hook being called by `deleteMany`.
While it's a genuine issue, it might qualify as a breaking change since some people may have empirically figured out that the wrong hook was called and decide to not implement the `afterDeleteMany` hook at all - though if you ask me, they deserve it for being so lazy 😛

alexandrebodin commented:
Hi @rgoupil I wanted to review one more time but the tests are not passing at the moment. Let me know if I can help :)

rgoupil commented:
Hmm, I'll investigate, curious I had no notification on the matter

rgoupil commented:
Rebased branch and fixed tests @alexandrebodin!

For some reason the tests are calling the `EntityManager` CRUD methods, such as `deleteMany` or `update`, without any parameters - sounds wrong but at least the tests are green now 😃

alexandrebodin commented on packages/core/database/lib/lifecycles/index.js:
```diff
@@ -30,33 +30,57 @@ const createLifecyclesProvider = db => {
+    /**
+     * @param {string} action
+     * @param {string} uid
+     * @param {{ param?: any, result?: any }} properties
```
```suggestion
     * @param {{ params?: any, result?: any }} properties
```

alexandrebodin commented on packages/core/database/lib/lifecycles/index.js:
```diff
@@ -30,33 +30,57 @@ const createLifecyclesProvider = db => {
+    /**
+     * @param {string} action
+     * @param {string} uid
+     * @param {{ param?: any, result?: any }} properties
```
```suggestion
     * @param {{ params?: any, result?: any }} properties
```

alexandrebodin reviewed (changes requested):
I tried to push on your branch but I think you disabled that when forking :) It's very minor but the rest looks good and works great!

rgoupil commented:
Oops, thanks for the suggestions @alexandrebodin, I just applied everything!

alexandrebodin reviewed (approved):
LGTM

Great work @rgoupil Thank you for your contribution 🔥
